### ISSUE TITLE
Automatic imports

### ISSUE THREAD

**Issue by pfitzseb** [julia-vscode/julia-vscode#1925]
It would be really nice if we could automatically add a `using Foo: bar` statement when completing `bar` anywhere in the editor, similar to what the TS LS does. I'm not sure how feasible this is though.

**Comment by ZacLN** [comment]
I'm not sure I completely understand what you mean - something similar to the "Fix missing reference" action?

**Comment by pfitzseb** [comment]
Well, we know what symbols are available in the current environment. So if I have e.g. `Plots` in my env, then autocompleting `plot` could add a `using Plots: plot` at the top of the file. This is complicated by the fact that imports aren't file specific in Julia, and that many packages could export/define the same binding and there's no way for us to choose the right one.

So overall this probably is very similar to the "Fix missing reference" action (which I haven't ever tried). Do we support that?

**Comment by ZacLN** [comment]
A bit similar - it turns a missing ref into a qualified name if possible (e.g. `argtail` -> `Base.argtail`) by searching over unexported names of packages that that have been used. 

This should be easy enough, we'd probably want to restrict it to completions where a few characters (3?) have already been typed otherwise you're going to get a ton of completions each time (quite costly). I'll have a play and stick up a PR

**Comment by BioTurboNick** [comment]
Did this recently get integrated? I keep finding random usings added to my code that I never intended.

**Comment by pfitzseb** [comment]
Yes. You can switch `julia.completionmode` to `exportedonly` to disable the feature (which does need some work).

**Comment by BioTurboNick** [comment]
Thanks! I mean it is a nice idea in general, I just find myself with unintentional autocompletes frequently enough. Maybe be better to have the using insertion behind an extra key press immediately after autocomplete occurs?

**Comment by scimas** [comment]
This should not be turned on by default since the extension is detecting even some of the Base types as "unexported symbols" and keeps adding import/using statements for them. Examples of such types are String, Bool.

**Comment by davidanthoff** [comment]
I feel that in general we should never modify the source code on completion in some other location than where the completion is happening, but instead have a feature like this as a code action that a user needs to opt-into with an extra click.

**Comment by pfitzseb** [comment]
Debatable. The TS extension does the same by default.

**Comment by davidanthoff** [comment]
> The TS extension does the same by default.

And it drives me nuts :) I think as a UI principle, things that automatically change things somewhere under the hood and I don't even see something happening in the UI are just weird. I think one of the nice things about completion is that the edits happen right in front of my eye, where I am focused right now, where my cursor is. I can't miss that. But something that edits parts of the doc that maybe aren't even visible right now seems like a step too much. I think we could have an option to enable it, but I would much prefer that the default was that a code action showed up for this.

**Comment by pfitzseb** [comment]
Fair enough. Let's disable it by default and slap an "Experimental" warning on the setting?

**Comment by lrnv** [comment]
What if I want the completion for everything but do not want the software to modify my usings statements while i'm not looking ? It's driving me nuts and polutes my git log...

**Comment by pfitzseb** [comment]
> Yes. You can switch `julia.completionmode` to `exportedonly` to disable the feature (which does need some work).

**Comment by lrnv** [comment]
This is not what the 'details' in the parameters says. 'Show completion for the current namespace' and 'Show completion from everything + automatic usings' are avaliable. I want 'Show completion for everything and never touch my code unles i said so' :)
